File: 0x17-doubly_linked_lists/palindromes.c
#include <stdio.h>
#include <stdbool.h>

bool is_palindrome(int n) {
    int num = n;
    int reverse = 0;
    while (num > 0) {
        reverse = reverse * 10 + num % 10;
        num /= 10;
    }
    return n == reverse;
}

int main() {
    int largest_palindrome = 0;
    for (int i = 100; i < 1000; i++) {
        for (int j = i; j < 1000; j++) {
            int product = i * j;
            if (is_palindrome(product) && product > largest_palindrome) {
                largest_palindrome = product;
            }
        }
    }
    printf("%d\n", largest_palindrome);
    return 0;
}

File: 0x17-doubly_linked_lists/palindromes.py
#!/usr/bin/python3
def is_palindrome(n):
    return str(n) == str(n)[::-1]
largest_palindrome = 0
# Start from the largest possible product and work downwards
for i in range(999, 99, -1):
    # Start from i to avoid redundant checks
    for j in range(i, 99, -1):
        product = i * j
        if product <= largest_palindrome:
            break  # No need to continue inner loop if product is smaller
        if is_palindrome(product):
            largest_palindrome = product
print(largest_palindrome)
